src/legacy/ui/public/vis/editors/default/components/default_editor_agg_group.tsx

```diff
@@ -0,0 +1,196 @@
+ * under the License.
+ */
+
+import React, { useEffect, useReducer } from 'react';
+import {
+  EuiTitle,
+  EuiDragDropContext,
+  EuiDroppable,
+  EuiDraggable,
+  EuiSpacer,
+  EuiPanel,
+} from '@elastic/eui';
+
+import { AggConfig } from '../../../agg_config';
+import { aggGroupNamesMap, AggGroupNames } from '../agg_groups';
+import { DefaultEditorAgg } from './default_editor_agg';
+import { DefaultEditorAggAdd } from './default_editor_agg_add';
+import { DefaultEditorAggCommonProps } from './default_editor_agg_common_props';
+import {
+  isInvalidAggsTouched,
+  isAggRemovable,
+  calcAggIsTooLow,
+} from './default_editor_agg_group_helper';
+import { aggGroupReducer, initAggsState, AGGS_ACTION_KEYS } from './default_editor_agg_group_state';
+import { Schema } from '../schemas';
+
+interface DefaultEditorAggGroupProps extends DefaultEditorAggCommonProps {
+  schemas: Schema[];
+  addSchema: (schems: Schema) => void;
+  reorderAggs: (group: AggConfig[]) => void;
+}
+
+function DefaultEditorAggGroup({
+  formIsTouched,
+  groupName,
+  lastParentPipelineAggTitle,
+  metricAggs,
+  state,
+  schemas = [],
+  addSchema,
+  onAggParamsChange,
+  onAggTypeChange,
+  onToggleEnableAgg,
+  removeAgg,
+  reorderAggs,
+  setTouched,
+  setValidity,
+}: DefaultEditorAggGroupProps) {
+  const groupNameLabel = aggGroupNamesMap()[groupName];
+  // e.g. buckets can have no aggs
+  const group: AggConfig[] = state.aggs.bySchemaGroup[groupName] || [];
+
+  const stats = {
+    max: 0,
+    count: group.length,
+    deprecate: false,
+  };
+
+  schemas.forEach((schema: Schema) => {
+    stats.max += schema.max;
+    stats.deprecate = schema.deprecate;
+  });
+
+  const [aggsState, setAggsState] = useReducer(aggGroupReducer, group, initAggsState);
+
+  const isGroupValid = Object.values(aggsState).every(item => item.valid);
+  const isAllAggsTouched = isInvalidAggsTouched(aggsState);
+
+  useEffect(() => {
+    // when isAllAggsTouched is true, it means that all invalid aggs are touched and we will set ngModel's touched to true
+    // which indicates that Apply button can be changed to Error button (when all invalid ngModels are touched)
+    setTouched(isAllAggsTouched);
+  }, [isAllAggsTouched]);
+
+  useEffect(() => {
+    // when not all invalid aggs are touched and formIsTouched becomes true, it means that Apply button was clicked.
+    // and in such case we set touched state to true for all aggs
+    if (formIsTouched && !isAllAggsTouched) {
+      Object.keys(aggsState).map(([aggId]) => {
+        setAggsState({
+          type: AGGS_ACTION_KEYS.TOUCHED,
+          payload: true,
+          aggId: Number(aggId),
+        });
+      });
+    }
+  }, [formIsTouched]);
+
+  useEffect(() => {
+    setValidity(isGroupValid);
+  }, [isGroupValid]);
+
+  interface DragDropProps {
+    result: { index: number };
+    provided: { index: number };
+  }
+  const onDragEnd = ({ result, provided }: DragDropProps) => {
```

Comment:
This shouldn't be functioning,

`result` and `provided` are separate arguments i.e. `(result, provided)` instead of `({ result, provided })`

`result` doesn't have an `index` property on it, instead there is `source.index` and `destination.index` (destination could also be null or undefined).

`provided` has the shape:

```
{
  announce: (message: string) => void
}
```

Getting the correct shapes here should remove the need for the `@ts-ignore` below

Reply:
That's my bad. Thank you for noticing and explanation.